Source/WebCore/platform/graphics/Damage.h

```diff
@@ -206,17 +206,52 @@ class Damage {
         return add(enclosingIntRect(rect));
     }
 
-    ALWAYS_INLINE bool add(const Damage& other)
+    ALWAYS_INLINE bool add(const Rects& rects)
     {
-        if (other.isEmpty() || !shouldAdd())
+        if (rects.isEmpty() || !shouldAdd())
             return false;
 
+        // When adding rects to an empty Damage and we know we will need to unite,
+        // we can unite the rects directly.
+        if (m_mode == Mode::Rectangles && m_rects.isEmpty()) {
+            auto rectsCount = rects.size();
+            auto gridArea = m_gridCells.unclampedArea();
+
+            if (rectsCount > gridArea) {
+                m_rects.grow(gridArea);
+                for (const auto& rect : rects) {
+                    if (rect.isEmpty())
+                        continue;
+
+                    m_minimumBoundingRectangle.unite(rect);
+                    unite(rect);
+                }
+
+                if (m_minimumBoundingRectangle.isEmpty()) {
+                    // All rectangles were empty.
+                    m_rects.clear();
+                    return false;
+                }
+                m_shouldUnite = true;
+
+                return true;
+            }
+        }
+
         bool returnValue = false;
-        for (const auto& rect : other.rects())
+        for (const auto& rect : rects)
             returnValue |= add(rect);
         return returnValue;
     }
 
+    ALWAYS_INLINE bool add(const Damage& other)
+    {
+        if (other.isEmpty() || !shouldAdd())
+            return false;
+
```

Comment:
A note for the future.

Here, if `*this` and `other` both have the same `m_rect` and both `m_shouldUnite==true`, we can add another fastpath that would just iterate the `i` from `0` to `m_rects.size()` and just `m_rects[i].unite(other.m_rects[i])` as in that case both damage objects have the same layout of grid-powered rectangles.

The above fastpath would be useful in case of `AcceleratedSurfaceDMABuf::RenderTarget::addDamage` where usually all the conditions are met

Reply:
Let's do it in this patch then. Thanks!